lib/semantics/pointer-assignment.cc

```diff
@@ -322,29 +345,93 @@ parser::Message *PointerAssignmentChecker::Say(A &&... x) {
   return msg;
 }
 
+static bool CheckPointerBounds(
+    evaluate::FoldingContext &context, const PointerAssignment &assignment) {
+  auto &messages{context.messages()};
+  const SomeExpr &lhs{assignment.lhs};
+  const SomeExpr &rhs{assignment.rhs};
+  bool isBoundsRemapping{false};
+  std::size_t numBounds{std::visit(
+      common::visitors{
+          [&](const PointerAssignment::BoundsSpec &bounds) {
+            return bounds.size();
+          },
+          [&](const PointerAssignment::BoundsRemapping &bounds) {
+            isBoundsRemapping = true;
+            evaluate::ExtentExpr lhsSizeExpr{1};
+            for (const auto &bound : bounds) {
+              lhsSizeExpr = std::move(lhsSizeExpr) *
+                  (common::Clone(bound.second) - common::Clone(bound.first) +
+                      evaluate::ExtentExpr{1});
+            }
+            if (std::optional<std::int64_t> lhsSize{evaluate::ToInt64(
+                    evaluate::Fold(context, std::move(lhsSizeExpr)))}) {
+              if (auto shape{evaluate::GetShape(context, rhs)}) {
+                if (std::optional<std::int64_t> rhsSize{
+                        evaluate::ToInt64(evaluate::Fold(
+                            context, evaluate::GetSize(std::move(*shape))))}) {
+                  if (*lhsSize > *rhsSize) {
+                    messages.Say(
+                        "Pointer bounds require %d elements but target has"
+                        " only %d"_err_en_US,
+                        *lhsSize, *rhsSize);  // 10.2.2.3(9)
+                  }
+                }
+              }
+            }
+            return bounds.size();
+          },
+      },
+      assignment.bounds)};
+  if (numBounds > 0) {
+    if (lhs.Rank() != static_cast<int>(numBounds)) {
+      messages.Say("Pointer '%s' has rank %d but the number of bounds specified"
+                   " is %d"_err_en_US,
+          lhs.AsFortran(), lhs.Rank(), numBounds);  // C1018
+    }
+  }
+  if (isBoundsRemapping && rhs.Rank() > 1 &&
```

Comment:
Should it be `rhs.Rank() != 1` instead or is the case `rhs.Rank() == 0` already handled ?

Reply:
Currently when the rhs is a scalar, `IsSimplyContiguous` returns true. That seems reasonable, though I'm not positive it is correct.

I'll change this to `!= 1` because it more correctly expresses the check. And if we decide a scalar is not simply contiguous and change the function, this will be doing the right thing.

Reply:
The *simply contiguous* property is defined in the standard only as an attribute of arrays, but it seemed to me to be digging a pitfall to not further define it vacuously true for scalars.